Prevent an error in `<ServerSideRender>` by allowing `POST` request

kienstra opened the pull request:
## Description
Allows optionally using `POST` requests in `<ServerSideRender>` to prevent an error when the `attributes` object is too big.

`<ServerSideRender>` uses GET requests [in apiFetch()](https://github.com/WordPress/gutenberg/blob/b16542e1ac508a6d31f1c87680aa73b1af5db3bf/packages/server-side-render/src/server-side-render.js#L61), but they can fail when the `attributes` object is too big:

![error-att](https://user-images.githubusercontent.com/4063887/77268860-dc820e00-6c6c-11ea-8b65-edd2887ce465.png)

This uses @swissspidy's suggestion in https://github.com/WordPress/gutenberg/issues/16396#issuecomment-508709339 to allow using `POST` requests for large attributes. 

It adds an optional prop `httpMethod`, and if it's `'POST'`, it causes `<ServerSideRender>` to make `POST` requests. It's prevented errors in my testing.

Also, it depends on https://github.com/WordPress/wordpress-develop/pull/196.

Reported in:
https://github.com/WordPress/gutenberg/issues/19935
https://github.com/WordPress/gutenberg/issues/16396#issuecomment-508709339
https://wordpress.org/support/topic/error-loading-block-the-response-is-not-a-valid-json-response-2/

## How has this been tested?

1. `wp core update --version=5.5-RC1 ` # or any version 5.5+, as this depends on https://github.com/WordPress/wordpress-develop/pull/196, which was merged into 5.5
2. `checkout` this PR's branch and do `npm install && npm run dev`
3. clone https://github.com/kienstra/ssr-demo into your `plugins/` directory in order to test this PR
4. Follow the [Usage](https://github.com/kienstra/ssr-demo#usage) instructions to test this PR with the `httpMethod` prop 
5. Expected: even a very large `attributes` object doesn't cause an error in the `<ServerSideRender>` component
6. Actual: same as expected
7. Regression testing on Core blocks that use `<ServerSideRender>` looks good, like 'Archives', 'Latest Comments', and 'Latest Posts'

## Screenshots <!-- if applicable -->
![expected-block](https://user-images.githubusercontent.com/4063887/77268951-2cf96b80-6c6d-11ea-97e0-eec37e24b785.png)

## Types of changes
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

kienstra committed:
Add a prop to make POST instead of GET requests

There can be an issue if the attributes object is big,
and the GET request fails and causes an error message
in the block.
This allows POSTing the attributes, preventing the error message.
See the PR to Core supporting the POST request.

kienstra committed:
Empty commit to trigger Travis build

There are 2 failed e2e tests,
though it doesn't look like they're caused by this PR.

kienstra committed:
Another empty commit to trigger Travis build

kienstra commented:
The E2E test failures above don't look to be related to this PR, but maybe I missed something.

kienstra committed:
Another empty commit to trigger build

kienstra committed:
Pass attributes in an object, as that's what Core now accepts

Now, Core accepts POST requests, with the attributes in an array.
https://github.com/WordPress/wordpress-develop/commit/5460e0df1e9a8fabecb44cf162478c70efa44ce1

kienstra committed:
Merge branch 'master' into add/ssr-post-support

kienstra commented:
Now that https://github.com/WordPress/wordpress-develop/commit/5460e0df1e9a8fabecb44cf162478c70efa44ce1 was merged, the dependency for this PR is satisfied.

kienstra committed:
Merge branch 'master' into add/ssr-post-support

kienstra committed:
Empty commit to trigger a build

There was a failed e2e test:
https://github.com/WordPress/gutenberg/pull/21068/checks?check_run_id=925354338#step:7:15
This doesn't look related to this PR,
but I don't know

kienstra commented:
Hi @ocean90,
Hope you're doing well.

Would you be able to review this, or point me to someone who could?

The [core patch](https://github.com/WordPress/wordpress-develop/pull/196) for this is merged. It now has the latest from `master`, and still works.

Thanks, Dominik!

swissspidy commented on packages/server-side-render/README.md:
```diff
@@ -48,6 +48,12 @@ Examples: "my-custom-server-side-rendered".
 - Type: `String`
 - Required: No
 
+### requestBody
```
IMO `requestBody` is not a good name for this. It's not clear that this is a `bool` flag. I would expect `requestBody` to allow me to specify the actual request body.

Why not rename it to something like `httpMethod`? It would allow `GET` or `POST`.

Alternatively, `requestBody` could accept an object. If I pass `requestBody={{ attributes }}`, `method` and `path` would be handled accordingly,  and the `attributes` prop would be ignored.

kienstra replied:
Thanks, @swissspidy! I'll rename `requestBody` to `httpMethod`.

kienstra replied:
Thanks, that's a good suggestion. It's applied in https://github.com/WordPress/gutenberg/compare/6eb0b0803aa53ca6001f9afdc97f88857d364eab...42153114ea71e2690a83fdba1317f268a986e4ac

kienstra committed:
Change the prop to httpMethod, as it's more accurate

As Pascal mentioned,
it's easier to understand what this does.

kienstra committed:
Add a trailing comma, remve trailing whitespace

ocean90 commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -53,12 +53,27 @@ export class ServerSideRender extends Component {
+		const fetchRequest = ( this.currentFetchRequest = apiFetch( {
+			path,
+			data,
+			method: httpMethod,
```
I think we should make sure that only POST and GET are used.

```suggestion
			method: isPostRequest ? 'POST' : 'GET',
```

kienstra replied:
Good idea, I'll make that change.

kienstra replied:
Committed in https://github.com/WordPress/gutenberg/pull/21068/commits/6ce918abd8321e986c6bbea2e59325607f2fab40

ocean90 reviewed (changes requested):
Would be good to document this change in CHANGELOG.md.

kienstra committed:
Update packages/server-side-render/src/server-side-render.js

Commit Dominik's suggestion to only allow POST and GET requests

Co-authored-by: Dominik Schilling <dominikschilling+git@gmail.com>

kienstra committed:
Add a CHANGELOG.md entry for the new prop

Following https://github.com/WordPress/gutenberg/tree/9bebc902f0c9d2c6d9ec47411250ea20cf038966/packages#maintaining-changelogs

kienstra committed:
Add the protocol to the URL

kienstra committed:
Use markdown link syntax for the PR link

kienstra commented:
Hi @ocean90,
>Would be good to document this change in CHANGELOG.md.

Good idea, the `CHANGELOG.md` entry [is now added](https://github.com/WordPress/gutenberg/compare/6ce918abd8321e986c6bbea2e59325607f2fab40...7f695e78c3bba01f9b55390a0d775c950281007e).

JohnRDOrazio commented:
Was this patch not included in 5.5?

swissspidy commented:
@JohnRDOrazio No. As you can see this PR has not been merged yet.

JohnRDOrazio commented:
ok, is there a workaround I can use until it's merged? A filter I can use to use POST instead of GET for ServerSideRender?

kienstra commented:
There's not really a workaround, other than forking this component. It'd be great if this could be merged, I really appreciate the reviews here.

kienstra commented:
Thanks a lot for merging this!
